### ISSUE TITLE
PropertySetterFactory: don't cache property setters globally as this leads to problems in the transformer

### ISSUE THREAD

**Issue by tbosch** [angular/angular#2359]
Old (see comments below):
Was: Transformers: Don't set `reflector.reflectionCapabilities` over an async gap

In [generator.dart](https://github.com/angular/angular/blob/master/modules/angular2/src/transform/template_compiler/generator.dart#L101-L109) we first set `reflector.reflectionCapabilities`, then wait for the compiler to compile and then reset the state. Although the caller of this particular method sequences calls, the call sites of generator.dart don't do this. 

This yields to adding setters/getters to the wrong `ng_deps.dart` files.

Solution:

1. change `DomRenderCompiler` to have a new method `compileSync`, which takes a `ViewDefinition` and a template string. The normal `compile` method can delegate to this one internally.
2. change `generator.dart` to first load the template, then set a temporal `reflector.reflectionCapabilities`, call `compiler.compileSync` and reset the temporal `reflector.reflectionCapabilities`.

**Comment by tbosch** [comment]
@naomiblack I think we should to a patch release for this one, as otherwise transformers are not really usable in Google3...

**Comment by kegluneq** [comment]
Is it true that `getChangeDetectorDefinitions` makes the same requests to `ReflectionCapabilities` as `compile` does? 

This is what [this block of asserts](https://github.com/angular/angular/blob/master/modules/angular2/src/transform/template_compiler/generator.dart#L56) was meant to test and so far it's come back clean.

If so, we can use the `RecordingReflectionCapabilities` in that area to generate our getters, setters, & methods instead, avoiding the async gap.

**Comment by tbosch** [comment]
I don't think so: The compiler parses expressions and requests getters/setters/methods from the reflector. Also, it requests getters/setters for property bindings. @vsavkin what do you think?

**Comment by kegluneq** [comment]
A quick test confirms that this strategy won't work.

Another potential option is to use the data structures in the `ProtoViewDto` returned by `compile` to figure out which `getter`s, `setter`s, and `method`s we need to register. This seems a bit more error-prone than actually pulling them out of the `ReflectionCapabilities` after `compile`, but avoids playing with global state, which just smells bad.

**Comment by tbosch** [comment]
After talking to @kegluneq we should do the following:

Don't create a new method on the `DomCompiler` but use `CompilePipeline` directly. The long term plan would be switching back to using the `DomCompiler` and only relying on its result instead of listening to the calls to the reflector.

**Comment by tbosch** [comment]
The reason for this issue was that in some cases the transpiler did not generate all of the property setters when a `ngFor` directive was used. It turns out, that the transpiler code is correct (although the code passage in this issue seemed suspicious). I could verify this via `print` statements that I added into the transformer when it replaces the `ReflectionCapabilities`. 

The real problem is in the [property_setter_factory.ts](https://github.com/angular/angular/blob/master/modules/angular2/src/render/dom/view/property_setter_factory.ts#L16): This is caching property setters in a module global variable. Because of this, even though the transpiler recreated the compiler for every template it only got property setters for e.g. `ngForOf` only once accross all compiled files of all entry points. Not caching these property setters leads to correctly generated setters for every template by the transpiler (I checked the `.ng_deps.dart` files).

**Comment by tbosch** [comment]
@kegluneq I think we don't need to change the transformer. However, could you verify that Dart won't make calls to https://github.com/angular/angular/blob/master/modules/angular2/src/transform/template_compiler/transformer.dart#L30 in parallel, i.e. call it again while another call is still pending? If this is the case, the transformer code is good.

I will create a separate PR for changing the property_setter_factory.ts to not cache, and also add a test case for the transformer.

**Comment by kegluneq** [comment]
I'll do some digging on this today, but my current understanding is that even if this is working now there is no guarantee that it will work in the future. Using the pipeline directly might be a good way to go until we directly process the `ProtoViewDto` to determine what we need.

**Comment by kegluneq** [comment]
I confirmed that we should still make this change, even if it is not the cause of the behavior you were observing.

**Comment by tbosch** [comment]
Landed in master via d1b35f9174dbef808fdff655f9c4b42b1c125242 and 87b3b718e329ccf9cc720d2e0f1b5a10d611daaf
